tests/MongoDB.Driver.Legacy.Tests/MongoClientTests.cs

```diff
@@ -56,5 +60,99 @@ public void UsesSameMongoServerWhenReadPreferenceTagsAreTheSame()
+        {
+            var subject = CreateSubject();
+            var cluster = ClusterDescriptionParser.Parse(clusterDescriptionJson);
+            var connectedServers = cluster.Servers.Where(s => s.State == ServerState.Connected);
+
+            var result = subject.SelectServers(cluster, connectedServers);
+
+            AreSessionsSupportedServerSelectorReflector.ClusterDescription(subject).Should().BeSameAs(cluster);
+        }
+
+        [Theory]
+        [InlineData("{ connectionMode : 'Direct', clusterType : 'ReplicaSet', servers : [ { state : 'Connected', type : 'ReplicaSetArbiter' } ]}")]
+        public void SelectServers_should_return_all_servers_when_connection_mode_is_direct(string clusterDescriptionJson)
+        {
+            var subject = CreateSubject();
+            var cluster = ClusterDescriptionParser.Parse(clusterDescriptionJson);
+            var connectedServers = cluster.Servers.Where(s => s.State == ServerState.Connected).ToList();
+
+            var result = subject.SelectServers(cluster, connectedServers);
+
+            result.Should().BeSameAs(connectedServers);
+        }
+
+        [Theory]
+        [InlineData("{ clusterType : 'ReplicaSet', servers : [ { state : 'Connected', type : 'ReplicaSetArbiter' }, { state : 'Connected', type : 'ReplicaSetPrimary' } ]}")]
+        public void SelectServers_should_return_data_bearing_servers_when_connection_mode_is__not_direct(string clusterDescriptionJson)
+        {
+            var subject = CreateSubject();
+            var cluster = ClusterDescriptionParser.Parse(clusterDescriptionJson);
+            var connectedServers = cluster.Servers.Where(s => s.State == ServerState.Connected).ToList();
+            var dataBearingServers = connectedServers.Skip(1).Take(1);
+
+            var result = subject.SelectServers(cluster, connectedServers);
+
+            result.Should().Equal(dataBearingServers);
+        }
+
+        // private methods
+        private IServerSelector CreateSubject()
+        {
+            return AreSessionsSupportedServerSelectorReflector.CreateInstance();
+        }
+    }
+
+    public static class MongoClientReflector
+    {
+        public static bool? AreSessionsSupported(this MongoClient obj, ClusterDescription clusterDescription) => (bool?)Reflector.Invoke(obj, nameof(AreSessionsSupported), clusterDescription);
+    }
+
+    public static class AreSessionsSupportedServerSelectorReflector
+    {
+        public static IServerSelector CreateInstance()
+        {
+            var type = typeof(MongoClient).GetTypeInfo().Assembly.GetType("MongoDB.Driver.MongoClient+AreSessionsSupportedServerSelector");
+            return (IServerSelector)Activator.CreateInstance(type);
+        }
+
+        public static ClusterDescription ClusterDescription(IServerSelector obj) => (ClusterDescription)Reflector.GetFieldValue(obj, nameof(ClusterDescription), BindingFlags.Public | BindingFlags.Instance);
```

Comment:
We can't use the usual extension method syntax here because the AreSessionsSupportedServerSelector class is private so it's not possible to define an extension method for it.

We could (in theory) define an extension method on IServerSelector, but that would be dangerous because we don't want such an extension method to be accidentally used with an instance of IServerSelector that is not an instance of AreSessionsSupportedServerSelector.